File: examples/legacy-architecture-samples/simple-stack-example-mvvm-fragments/src/main/java/com/zhuinden/simplestackexamplemvvm/application/MainActivity.java
package com.zhuinden.simplestackexamplemvvm.application;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.navigation.NavigationView;
import androidx.fragment.app.Fragment;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.zhuinden.simplestack.Backstack;
import com.zhuinden.simplestack.History;
import com.zhuinden.simplestack.StateChange;
import com.zhuinden.simplestack.StateChanger;
import com.zhuinden.simplestack.navigator.Navigator;
import com.zhuinden.simplestackexamplemvvm.R;
import com.zhuinden.simplestackexamplemvvm.application.injection.Injector;
import com.zhuinden.simplestackexamplemvvm.features.statistics.StatisticsKey;
import com.zhuinden.simplestackexamplemvvm.features.tasks.TasksKey;
import com.zhuinden.simplestackexamplemvvm.util.ServiceProvider;

import butterknife.BindView;
import butterknife.ButterKnife;

import static androidx.drawerlayout.widget.DrawerLayout.LOCK_MODE_LOCKED_CLOSED;
import static androidx.drawerlayout.widget.DrawerLayout.LOCK_MODE_UNLOCKED;

/**
 * Created by Zhuinden on 2017.07.26..
 */

public class MainActivity
        extends AppCompatActivity
        implements StateChanger {
    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.fab)
    FloatingActionButton fab;

    @BindView(R.id.drawer_layout)
    DrawerLayout drawerLayout;

    @BindView(R.id.nav_view)
    NavigationView navigationView;

    @BindView(R.id.contentFrame)
    ViewGroup contentFrame;

    private FragmentStateChanger fragmentStateChanger;
    private ActionBarDrawerToggle drawerToggle;

    private final NavigationView.OnNavigationItemSelectedListener navigationItemSelectedListener = (NavigationView.OnNavigationItemSelectedListener) item -> {
        switch(item.getItemId()) {
            case R.id.list_navigation_menu_item:
                Navigator.getBackstack(this).goTo(TasksKey.create());
                break;
            case R.id.statistics_navigation_menu_item:
                Navigator.getBackstack(this).goTo(StatisticsKey.create());
            default:
                break;
        }
        setCheckedItem(item.getItemId());
        // Close the navigation drawer when an item is selected.
        drawerLayout.closeDrawers();
        return true;
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.main_activity);

        ButterKnife.bind(this);

        setupDrawer();

        fragmentStateChanger = new FragmentStateChanger(this, getSupportFragmentManager(), R.id.contentFrame);

        Backstack backstack = Navigator.configure()
                .setStateChanger(this)
                .setScopedServices(new ServiceProvider())
                .setDeferredInitialization(true)
                .install(this, contentFrame, History.of(TasksKey.create()));

        BackstackHolder backstackHolder = Injector.get().backstackHolder();
        backstackHolder.setBackstack(backstack); // <-- make Backstack globally available through Dagger, singleInstance only!

        Navigator.executeDeferredInitialization(this);
    }

    @Override
    public void onBackPressed() {
        if(!Navigator.onBackPressed(this)) {
            super.onBackPressed();
        }
    }

    private void setupDrawer() {
        drawerLayout.setStatusBarBackground(R.color.colorPrimaryDark);
        navigationView.setNavigationItemSelectedListener(navigationItemSelectedListener);

        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        drawerToggle = new ActionBarDrawerToggle(this, drawerLayout, toolbar, R.string.open, R.string.close) {
            @Override
            public void onDrawerClosed(View drawerView) {
                super.onDrawerClosed(drawerView);
                supportInvalidateOptionsMenu();
            }

            @Override
            public void onDrawerOpened(View drawerView) {
                super.onDrawerOpened(drawerView);
                supportInvalidateOptionsMenu();
            }
        };
        // noinspection deprecation
        drawerLayout.setDrawerListener(drawerToggle);

        drawerToggle.setToolbarNavigationClickListener(v -> Navigator.onBackPressed(this));
        actionBar.setDisplayHomeAsUpEnabled(false);
        actionBar.setHomeButtonEnabled(true);
    }

    private void setCheckedItem(int navigationItemId) {
        Menu menu = navigationView.getMenu();
        for(int i = 0; i < menu.size(); i++) {
            MenuItem menuItem = menu.getItem(i);
            menuItem.setChecked(menuItem.getItemId() == navigationItemId);
        }
    }

    public void setupViewsForKey(BaseKey key) {
        if(key.shouldShowUp()) {
            drawerLayout.setDrawerLockMode(LOCK_MODE_LOCKED_CLOSED, GravityCompat.START);
            drawerToggle.setDrawerIndicatorEnabled(false);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        } else {
            drawerLayout.setDrawerLockMode(LOCK_MODE_UNLOCKED, GravityCompat.START);
            getSupportActionBar().setDisplayHomeAsUpEnabled(false);
            drawerToggle.setDrawerIndicatorEnabled(true);
        }
        drawerToggle.syncState();
        setCheckedItem(key.navigationViewId());
        supportInvalidateOptionsMenu();
        if(key.isFabVisible()) {
            fab.show();
        } else {
            fab.hide();
        }
        Fragment fragment = getSupportFragmentManager().findFragmentByTag(key.getFragmentTag());
        key.setupFab(fragment, fab);
    }

    @Override
    public void handleStateChange(@NonNull StateChange stateChange, @NonNull Callback completionCallback) {
        if(!stateChange.isTopNewKeyEqualToPrevious()) {
            fragmentStateChanger.handleStateChange(stateChange);
            setupViewsForKey(stateChange.topNewKey());
            String title = stateChange.<BaseKey>topNewKey().title(getResources());
            setTitle(title == null ? getString(R.string.app_name) : title);
        }
        completionCallback.stateChangeComplete();
    }
}

File: samples/advanced-samples/mvvm-sample/src/main/java/com/zhuinden/simplestackexamplemvvm/application/MainActivity.kt
package com.zhuinden.simplestackexamplemvvm.application


import android.os.Bundle
import android.view.MenuItem
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import com.zhuinden.liveevent.observe
import com.zhuinden.simplestack.History
import com.zhuinden.simplestack.SimpleStateChanger
import com.zhuinden.simplestack.StateChange
import com.zhuinden.simplestack.navigator.Navigator
import com.zhuinden.simplestackexamplemvvm.R
import com.zhuinden.simplestackexamplemvvm.databinding.MainActivityBinding
import com.zhuinden.simplestackexamplemvvm.features.statistics.StatisticsKey
import com.zhuinden.simplestackexamplemvvm.features.tasks.TasksKey
import com.zhuinden.simplestackexamplemvvm.util.showSnackbar
import com.zhuinden.simplestackextensions.fragments.DefaultFragmentStateChanger
import com.zhuinden.simplestackextensions.navigatorktx.backstack
import com.zhuinden.simplestackextensions.services.DefaultServiceProvider
import com.zhuinden.simplestackextensions.servicesktx.get

class MainActivity : AppCompatActivity(), SimpleStateChanger.NavigationHandler {
    private lateinit var fragmentStateChanger: DefaultFragmentStateChanger
    private lateinit var snackbarTextEmitter: SnackbarTextEmitter
    private lateinit var binding: MainActivityBinding

    @Suppress("DEPRECATION")
    private val backPressedCallback = object : OnBackPressedCallback(true) {
        override fun handleOnBackPressed() {
            if (!Navigator.onBackPressed(this@MainActivity)) {
                this.remove()
                onBackPressed()  // this is the reliable way to handle back for now
                this@MainActivity.onBackPressedDispatcher.addCallback(this)
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = MainActivityBinding.inflate(layoutInflater)
        setContentView(binding.root)

        onBackPressedDispatcher.addCallback(backPressedCallback) // this is the reliable way to handle back for now

        val app = application as CustomApplication
        val globalServices = app.globalServices

        snackbarTextEmitter = globalServices.get()

        binding.drawerLayout.setStatusBarBackground(R.color.colorPrimaryDark)

        binding.navView.setNavigationItemSelectedListener { item: MenuItem ->
            when (item.itemId) {
                R.id.list_navigation_menu_item -> backstack.goTo(TasksKey)
                R.id.statistics_navigation_menu_item -> backstack.goTo(StatisticsKey)
                else -> {
                }
            }

            setCheckedItem(item.itemId)

            binding.drawerLayout.closeDrawers()
            true
        }

        fragmentStateChanger = DefaultFragmentStateChanger(supportFragmentManager, R.id.contentFrame)

        snackbarTextEmitter.snackbarText.observe(this) { textRes ->
            showSnackbar(binding.root, getString(textRes))
        }

        Navigator.configure()
            .setStateChanger(SimpleStateChanger(this))
            .setGlobalServices(globalServices)
            .setScopedServices(DefaultServiceProvider())
            .install(this, binding.contentFrame, History.of(TasksKey))
    }

    fun toggleLeftDrawer() {
        val drawerLayout = binding.drawerLayout

        if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
            drawerLayout.closeDrawer(GravityCompat.START)
        } else {
            drawerLayout.openDrawer(GravityCompat.START)
        }
    }

    private fun setCheckedItem(navigationItemId: Int) {
        val menu = binding.navView.menu
        for (i in 0 until menu.size()) {
            val menuItem = menu.getItem(i)
            menuItem.isChecked = menuItem.itemId == navigationItemId
        }
    }

    override fun onNavigationEvent(stateChange: StateChange) {
        binding.drawerLayout.setDrawerLockMode(when {
            stateChange.getNewKeys<BaseKey>().size > 1 -> DrawerLayout.LOCK_MODE_LOCKED_CLOSED
            else -> DrawerLayout.LOCK_MODE_UNLOCKED
        }, GravityCompat.START)

        val topNewKey = stateChange.topNewKey<BaseKey>()

        if (topNewKey is TasksKey) {
            setCheckedItem(R.id.list_navigation_menu_item)
        }

        if (topNewKey is StatisticsKey) {
            setCheckedItem(R.id.statistics_navigation_menu_item)
        }

        fragmentStateChanger.handleStateChange(stateChange)
    }
}